### DISCUSSION TITLE
Set a default value for component key when not specified?

### DISCUSSION THREAD

**Discussion by gsxdsm** [tetra-framework/tetra#86] [General]
I think we should set a default key for each component - right now when it's not specified by the user in the tag, we run into issues re-rendering a component that has child state.

Will create a pull request with proposed change but wanted to check to see if this was intentional.

```html
        <div>
           <div class="input-group mb-2">
               <input type="text" x-model="name" class="form-control"
                   placeholder="New job..." @keyup.enter="add_job(name)">
               <button class="btn btn-primary" @click="add_job(name)">Add</button>
               <button class="btn" @click="name=''">Clear</button>
           </div>
           <div class="list-group">
               {% for job in jobs %}
                {% @ JobItem job=job key=job.id / %}
               {% endfor %}
           </div>
           {% @ Counter / %}
        </div>
```

The Counter component has no key, but when there is component state (after first load and an action fires) it will error out. By setting a default key when non exists (just to the component class name) it works as expected.

**Comment by gsxdsm** [comment]
@nerdoc for your input (this came up as I was going through testing components as directories and forms)

**Reply by nerdoc** [reply]
Good catch. I thought about that months ago already. But never had time to implement it. But it was another bug  on my (huge) TODO list.
I even thought about giving each component instance a distinct key automatically, so more than one components could coexist on the page even without addressing a key to them.
But using the component full name as key at least fixes this big (I maybe introduced with dir components?).
Thanks a lot!

**Comment by gsxdsm** [comment]
Awesome. Agree a distinct key would be even better. I was thinking about
this briefly but didn't have any good option for something that was stable.
The closest thing was trying to use the dom path as a key but that breaks
if things move around the page. What were you thinking? I'll thinking about
it more after the form file bits are done.

**Reply by nerdoc** [reply]
Yes, my initial idea was an UUID, but that's of course nonsense, when regenerated at each request. What maybe would work is creating an UUID at "the beginning", and saving it / including it in the (encrypted) state.
This has another security drawback, as it would then be available in the encrypted state, and in freetext, so decrypting would be easier for a attacker. But this is true for all other attrs (which are in the encrypted state AND in clear text as well).
I'm no security expert here. My guts tell me that it would be ok, as the rest of the attrs do it the same way.
